File: java/leetcode/dp/combination_sum_iv.java
/**
 * 377. Combination Sum IV
 * Given an integer array with all positive numbers and no duplicates, find the number of possible combinations that add up to a positive integer target.

Example:

nums = [1, 2, 3]
target = 4

The possible combination ways are:
(1, 1, 1, 1)
(1, 1, 2)
(1, 2, 1)
(1, 3)
(2, 1, 1)
(2, 2)
(3, 1)

Note that different sequences are counted as different combinations.

Therefore the output is 7.
 

Follow up:
What if negative numbers are allowed in the given array?
How does it change the problem?
What limitation we need to add to the question to allow negative numbers?
 */

import java.util.Arrays;

public class combination_sum_iv {

    public int combinationSum4(int[] nums, int target) {
        if (nums == null || nums.length == 0) {
            return 0;
        }
        int[] maximum = new int[target+1];
        maximum[0] = 1;
        Arrays.sort(nums);
        for (int i = 1; i <= target; i++) {
            for (int j = 0; j < nums.length && nums[j] <= i; j++) {
                maximum[i] += maximum[i-nums[j]];
            } 
        }
        return maximum[target];
    }
}

File: golang/leetcode/dp/combination_sum_iv.go
package dp

import "sort"

// 377. Combination Sum IV

func combinationSum4(nums []int, target int) int {
	if len(nums) == 0 {
		return 0
	}
	dp := make([]int, target+1)
	dp[0] = 1
	sort.Ints(nums)
	for i := 1; i <= target; i++ {
		for j := 0; j < len(nums) && nums[j] <= i; j++ {
			dp[i] += dp[i-nums[j]]
		}
	}
	return dp[target]
}
